src/netdev_flow.c

```diff
@@ -0,0 +1,1063 @@
+        patts[index].spec = &udp_spec;
+        patts[index].mask = &udp_mask;
+        break;
+    default:
+        RTE_LOG(WARNING, NETDEV, "[%s]: unknown l4 proto\n", __func__);
+        break;
+    }
+
+    get_next_patts_index(index);
+    patts[index].type = RTE_FLOW_ITEM_TYPE_END;
+}
+
+/*
+ * @brief add l3/l4 filter actions
+ *
+ * @return void
+ */
+static inline void
+netdev_flow_add_filter_actions(struct netdev_flow *netdev_flow,
+                               struct rte_flow_action acts[])
+{
+    int index = 0;
+    struct netif_port *port;
+    static struct rte_flow_action_queue queue = { .index = 0 };
+
+    port = netif_port_get(netdev_flow->port_id);
+    if (unlikely(port == NULL)) {
+        RTE_LOG(ERR, NETDEV,
+                "[%s]: invalid netdev info on port\n",
+                __func__);
+        return;
+    }
+
+    queue = (struct rte_flow_action_queue) {
+        .index = netdev_flow->data.filter_info.queue_id,
+    };
+
+    /* queue action is essential */
+    acts[index].type = RTE_FLOW_ACTION_TYPE_QUEUE;
+    acts[index].conf = &queue;
+
+    /*
+     * attach an integer value to packets and
+     * set PKT_RX_FDIR and PKT_RX_FDIR_ID mbuf flags
+     */
+    netdev_flow_add_mark_action(port, acts, &index);
+
+    /* count action supported by mellanox only */
+    netdev_flow_add_count_action(port, acts, &index);
+
+    get_next_acts_index(index);
+    acts[index].type = RTE_FLOW_ACTION_TYPE_END;
+}
+
+/*
+ * @brief add egress flow attr
+ *
+ * @return void
+ */
+static inline void
+netdev_flow_add_egress_attribute(struct rte_flow_attr *attr)
+{
+    attr->egress = 1;
+}
+
+/*
+ * @brief add netdev flow on port
+ *
+ * @param port - dpdk or other type port
+ * @param netdev_flow - flow store on netif_port
+ *
+ * @return EDPVS_OK on success, EDPVS_DPDKAPIFAIL on failure
+ */
+int netdev_flow_add(struct netif_port *port,
+                    struct netdev_flow *netdev_flow)
+{
+    int err = EDPVS_OK;
+    struct rte_flow_error error;
+    struct rte_flow_attr attr;
+    struct rte_flow_item patts[DEFAULT_MAX_PATTERNS];
+    struct rte_flow_action acts[DEFAULT_MAX_ACTIONS];
+    union netdev_flow_query query;
+    struct rte_flow *flow = NULL;
+    portid_t port_id;
+
+    if (unlikely(netdev_flow == NULL || port == NULL)) {
+        RTE_LOG(ERR, NETDEV,
+                "[%s]: invalid netdev flow info on port\n",
+                __func__);
+        return EDPVS_INVAL;
+    }
+
+    port_id = port->id;
+
+    memset(&error, 0, sizeof(error));
+    memset(&attr, 0, sizeof(attr));
+    memset(patts, 0, sizeof(patts));
+    memset(acts, 0, sizeof(acts));
+
+    rte_rwlock_write_lock(&port->dev_lock);
+
+    switch (netdev_flow->type) {
+    case NETDEV_FLOW_TYPE_RSS:
+        /* setup rss queues info */
+        netdev_flow_add_ingress_attribute(netdev_flow, &attr);
+        netdev_flow_add_rss_patterns(netdev_flow, patts);
+        netdev_flow_add_rss_actions(port_id, netdev_flow, acts);
+        break;
```

Comment:
Does ixgbe devices support rss flow type? How does its performance  compared to the global RSS?